### DISCUSSION TITLE
Ability to define the JSON Schema description on a class separately (or as a subpart) of the docstring

### DISCUSSION THREAD

**Discussion by smarie** [pydantic/pydantic#2937] [Feature Request]
Dear pydantic maintainers,

It is frequent that classes ending up in public JSON Schemas are also useful for the application, and therefore have some additional methods. These may have their own docstring, but a global class-level developer-oriented docstring is sometimes needed. In that situation, the same code element (the class docstring) is used both for public JSON schema description and for developer help/maintenance, which limits drastically what can be written here.

I was wondering if we could provide a mechanism to have both descriptions coexist nicely in the class definition. The only workarounds I found to date are:

 - creating two classes: one "public" and one "internal" and having to copy over the information or wrap the public object when received. Not very convenient..
 - (only if the type is embedded as a field of another broader type) to override the description using `Field(description=)` in the wrapping type(s).
 - dynamically modifying the docstring later in the code using `MyClass.__doc__ = """the public JSON Schema documentation"""`

`pydantic` could instead offer a class decorator to provide the public description. Below an example of an enum type exposed directly as a query parameter in FastAPI.

```python
@schema_description("""Represents the various possible foo.""")
class Foo(Enum):
    """An enum to represent the various possible foo.

   Each element provides methods to convert data to ...
   In addition a class method is provided to create a Foo programmatically from ...
    """
```

This decorator would add a `__pydantic__` attribute to the `Foo` class containing a dictionary with `{"description": "Represents the various possible foo."}`. When generating the json schema associated with this class, pydantic would use this description instead of the docstring.

Note that the example above does not extend `BaseModel` but the same need is also relevant for base models.

Alternatively to the decorator, we could support some kind of marker to use inside the docstring to declare the exposed description. At first I thought that this would be more convenient, but it isn't, as defining such a marker and having Sphinx or other api doc generator happy with it will be a pain.

What do you think ?

**Comment by jclerman** [comment]
I just opened a separate discussion on the same topic (before seeing this one).  My suggestion was to provide this functionality via a new attribute of the `Config` subclass (where title can already be set) - but some way to accomplish this would be nice.
